app/controllers/collections_controller.rb

```diff
@@ -87,4 +117,47 @@ def set_collection
   def collection_params
     params.require(:collection).permit(:title, :description, :image, :image_url, :public, {:keywords => []}, {:material_ids => []}, {:event_ids => []})
   end
+
+  # Filter collection items based on a type
+  def item_class
+    case params[:type]
+    when 'Event'
+      return Event if TeSS::Config.feature['events']
+    when 'Material'
+      return Material if TeSS::Config.feature['materials']
+    end
+
+    raise ActiveRecord::AccessDenied
+  end
+
+  # since we have not checked all items we have to do a little work
+  # to add and remove only those that were checked now.
+  # may be able to write this in a nicer way using the magic of rails associations
+  def update_collection_items!
+    # remove unselected ones, if any exist
+    # one query per item, due to callbacks.
+    CollectionItem.where(collection_id: @collection.id,
+                         resource_id: unselected_collection_item_ids,
+                         resource_type: item_class.name).destroy_all
+
+    # find out which ones to add
+    CollectionItem.create!((selected_collection_item_ids - existing_collection_item_ids).map do |id|
+      { resource_id: id, resource_type: item_class.name, collection_id: @collection.id }
+    end)
+    # the after_save callback will probably still add a bunch of activities, so we can't avoid individual queries just yet.
+  end
+
+  def selected_collection_item_ids
+    Set.new(params[:item_ids]&.map(&:to_i))
+  end
+
+  def unselected_collection_item_ids
+    Set.new(params[:reviewed_item_ids]&.map(&:to_i)) - selected_collection_item_ids
+  end
+
+  def existing_collection_item_ids
+    Set.new(CollectionItem.where(collection_id: @collection.id,
+                                 resource_id: selected_ids,
+                                 resource_type: item_class.name).pluck(:resource_id))
```

Comment:
Could this be simplified by just doing e.g. `@collection.event_ids = (selected + existing IDs?)`

Reply:
something like that, but not completely (we need to filter for the reviewed item ids only)

Reply:
that approach would also load _all_ existing entries from the database, which is not necessary. I rewrote to use bulk inserts and deletes.